Consider symlink when saving localconfig.php

fritzmg opened the pull request:
We are using [Deployer](https://deployer.org/) to deploy our Contao 4 instances. In the config, we have added

```
system/config/localconfig.php
``` 

to the `shared_files` so that any changes to the system config in the back end do not get lost after the next deployment. However, this does not actually work as Contao simply overwrites the symlink with a regular `localconfig.php` file when saving.

This PR changes that: it evaluates the `realpath` of an existing `localconfig.php` and uses that path as the new destination.

fritzmg committed:
consider symlinks when saving localconfig.php

leofeyer commented on core-bundle/src/Resources/contao/library/Contao/Config.php:
```diff
@@ -244,37 +247,48 @@ public function save()
 		if (\function_exists('opcache_invalidate'))
 		{
-			opcache_invalidate(TL_ROOT . '/system/config/localconfig.php', true);
+			opcache_invalidate($strDestination, true);
```
Are you sure that this is correct? OPcache internally uses the realpath instead of the symlink?

fritzmg replied:
OPcache uses the realpath of a file internally as far as I know. So in the end it should not matter which path you invalidate.

Toflar replied:
FYI: Yes it does use the realpath internally. Here's the piece of code responsible for that in case you care: https://github.com/php/php-src/blob/5fdabeb9695d008afcc967bd6e0c60e889fa6883/ext/opcache/ZendAccelerator.c#L1293

leofeyer committed:
Merge branch '4.4' into consider-symlinks-for-localconfig

aschempp commented on core-bundle/src/Resources/contao/library/Contao/Config.php:
```diff
@@ -244,37 +247,48 @@ public function save()
+		// Get the realpath in case it is a symlink
+		if ($fs->exists($strDestination))
+		{
+			$strDestination = realpath($strDestination);
```
Using `$strDestination = realpath($strDestination) ?: $strDestination` would save use the `$fs->exists()` call.

aschempp reviewed (commented):
Is there a reason to switch to the filesystem? I'd prefer not to mess with this as much as possible…?

fritzmg commented:
> Is there a reason to switch to the filesystem? I'd prefer not to mess with this as much as possible…?

Yes, `\Contao\Files` does not allow operations outside the Contao root.

fritzmg committed:
remove unnecessary file_exists call

`realpath` already checks for the existence of the file and returns `FALSE` in that case.

aschempp reviewed (commented):
Ok, so while we rewrite this, I wonder why we use `system/tmp` and not just `tmpfile`?

fritzmg committed:
fix typo

leofeyer reviewed (changes requested):
I would prefer to make this PR as minimally invasive as possible.

* Do not switch to Sf Filesystem unless technically required.
* Do not add PathUtil. We have only made that change in Contao 4.10.

fritzmg commented:
> * Do not switch to Sf Filesystem unless technically required.

It is required, because `\Contao\Files` does not allow operations outside the project root. I can change the PR, so that only the `rename` is done with Symfony Filesystem and everything else stays with `\Contao\Files`, but that seems kind of pointless to me.

> * Do not add PathUtil. We have only made that change in Contao 4.10.

I did not add `PathUtil`. It is already there in Contao 4.4.

fritzmg commented:
> Ok, so while we rewrite this, I wonder why we use `system/tmp` and not just `tmpfile`?

I think only @leofeyer can answer that - but I agree with @leofeyer that this PR should stay as minimally invasive as possible. Thus I do not want to change this behaviour as I am unsure about implications and side-effects.

leofeyer commented:
>  I can change the PR, so that only the rename is done with Symfony Filesystem and everything else stays with \Contao\Files, but that seems kind of pointless to me.

Agreed. If there is a technical reason, switching to Sf Filesystem is ok.

fritzmg committed:
add PR reference

leofeyer committed:
Remove PathUtil

fritzmg commented:
> * Do not add PathUtil. We have only made that change in Contao 4.10.

Just to re-iterate on that: it is already there in Contao 4.4:

* https://github.com/contao/contao/blob/8bde4f14db65ebc6d8708e6ead01a58abca63ae4/core-bundle/composer.json#L73
* https://github.com/contao/contao/blob/8bde4f14db65ebc6d8708e6ead01a58abca63ae4/core-bundle/src/DependencyInjection/Configuration.php#L163

And since it's already there, it is much safer to use it than not to use it.

leofeyer committed:
Re-add PathUtil

leofeyer commented:
> And since it's already there, it is much safer to use it than not to use it.

It is so awfully inconsistent to use it in one method and not in the other three. But for the sake of upstream mergeability, I have re-added it in 14a67e5f9bf724f1aed983559a21092bc6f35076.

leofeyer commented:
Thank you @fritzmg.
